oci/default.nix

```diff
@@ -0,0 +1,558 @@
+
+      # filter out empty layers, this can happen if a layer has no entries and
+      # all its dependencies are excluded
+      nonEmptyLayers = builtins.concatMap (layer:
+        lib.optionals (builtins.pathExists "${layer}/layer.tar") [ layer ]
+      ) resolvedLayers;
+
+      nonEmptyLayersMetadata = map readMetadata nonEmptyLayers;
+
+      # merge PATH in env
+      configEnv = config.Env or [];
+      newEnv =
+        if (builtins.any (lib.hasPrefix "PATH=") configEnv) then
+          map (env:
+            if (lib.hasPrefix "PATH=" env) then
+              builtins.replaceStrings ["$PATH"] [path] env
+            else
+              env
+          ) configEnv
+        else
+          configEnv ++ [ "PATH=${path}" ];
+
+      configJson = builtins.toJSON {
+        inherit architecture os;
+        created = "1970-01-01T00:00:00Z";
+        rootfs = {
+          type = "layers";
+          diff_ids = map (metadata: metadata.digest) nonEmptyLayersMetadata;
+        };
+        config = (spec.config or {}) // {
+          Env = newEnv;
+        };
+      };
+
+      configHash = builtins.hashString "sha256" configJson;
+
+      manifestJson = builtins.toJSON {
+        schemaVersion = 2;
+        config = {
+          mediaType = "application/vnd.oci.image.config.v1+json";
+          size = builtins.stringLength configJson;
+          digest = "sha256:" + configHash;
+        };
+        layers = nonEmptyLayersMetadata;
+      };
+
+      manifestHash = builtins.hashString "sha256" manifestJson;
+
+      metadataJson = builtins.toJSON ({
+        mediaType = "application/vnd.oci.image.manifest.v1+json";
+        size = builtins.stringLength manifestJson;
+        digest = "sha256:" + manifestHash;
+        platform = {
+          inherit architecture os;
+        };
+      } // lib.optionalAttrs (tag != null) {
+        annotations."org.opencontainers.image.ref.name" = tag;
+      });
+    in
+    runCommand name {
+      inherit configJson manifestJson metadataJson;
+      passAsFile = [ "configJson" "manifestJson" "metadataJson "];
+
+      passthru = {
+        inherit spec resolvedLayers;
+        layers = nonEmptyLayers;
+      };
+    } ''
+      mkdir -p $out
+      cp $configJsonPath $out/config.json
+      cp $manifestJsonPath $out/manifest.json
+      cp $metadataJsonPath $out/metadata.json
+
+      mkdir -p $out/blobs/sha256
+      ln -s $out/config.json $out/blobs/sha256/${configHash}
+      ln -s $out/manifest.json $out/blobs/sha256/${manifestHash}
+      ${lib.concatMapStringsSep "\n" (layer: ''
+        cp -r ${layer}/blobs $out
+      '') nonEmptyLayers}
+    '';
+
+  /* Creates an OCI image directory containing various manifests
+
+     This is usable with skopeo using the path oci:<path to result>:tag
+
+     Example:
+
+       skopeo inspect --config oci:result
+
+       skopeo copy oci:result:0.1.0 docker-daemon:myimage:0.1.0
+  */
+  makeImageDirectory = { name, manifests, passthru ? {} }:
```

Comment:
Supporting multiple manifests allows adding multi-arch images but that's pending.